features/storage/kvstore/KVStore.h

```
@@ -0,0 +1,199 @@
+     * @param[in]  key                  Key.
+     * @param[in]  final_data_size      Final value data size.
+     * @param[in]  create_flags         Flag mask.
+     *
+     * @returns 0 on success or a negative error code on failure
+     */
+    virtual int set_start(set_handle_t *handle, const char *key, size_t final_data_size, uint32_t create_flags);
+
+    /**
+     * @brief Add data to incremental KVStore set sequence.
+     *
+     * @param[in]  handle               Incremental set handle.
+     * @param[in]  value_data           value data to add.
+     * @param[in]  data_size            value data size.
+     *
+     * @returns 0 on success or a negative error code on failure
+     */
+    virtual int set_add_data(set_handle_t handle, const void *value_data, size_t data_size);
+
+    /**
+     * @brief Finalize an incremental KVStore set sequence.
+     *
+     * @param[in]  handle               Incremental set handle.
+     *
+     * @returns 0 on success or a negative error code on failure
+     */
+    virtual int set_finalize(set_handle_t handle);
+
+    /**
+     * @brief Start an iteration over KVStore keys.
+     *
+     * @param[out] it                   Returned iterator handle.
+     * @param[in]  prefix               Key prefix (null for all keys).
+     *
+     * @returns 0 on success or a negative error code on failure
+     */
+    virtual int iterator_open(iterator_t *it, const char *prefix = NULL);
+
+    /**
+     * @brief Get next key in iteration.
+     *
+     * @param[in]  it                   Iterator handle.
+     * @param[in]  key                  Buffer for returned key.
+     * @param[in]  key_size             Key buffer size.
+     * @param[out] actual_key_size      Actual key size.
+     *
+     * @returns 0 on success or a negative error code on failure
+     */
+    virtual int iterator_next(iterator_t it, char *key, size_t key_size, size_t *actual_key_size);
```

We should default `actual_key_size` to NULL (assuming NULL is supported)

---

Actually, I removed this parameter altogether (also from the confluence page). Key is a null terminated string, so actual key size is not required.
This got dismissed right at the time you added your comment. :)

---

Haha, fair enough

What happens if the name is larger than the buffer?

---

Should return an error (as truncating the string will return a wrong key). We have define for max key size, so user should be aware of it when allocating the buffer for it.

---

Good point, sounds like a plan to me